stdlib/toolchain/Compatibility51/ProtocolConformance.cpp

```diff
@@ -0,0 +1,677 @@
+//===--- ProtocolConformance.cpp - Swift conformance checking backport ----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Checking and caching of Swift protocol conformances.
+//
+// This is a version of the Swift 5.2 protocol conformance cache implementation
+// adapted for backporting to Swift 5.1 with the following fixes applied:
+//
+// - rdar://problem/59460603, fixing a problem where the conformance cache would
+//   eagerly instantiate metadata for types when not necessary, causing crashes
+//   if instantiating the type relied on weak-linked symbols that aren't
+//   available on the client OS
+//
```

Comment:
I suggest mentioning that this also gets used in the 5.0 library, so future maintainers know that they need to be compatible with both.